gapis/resolve/dependencygraph2/dce.go

```diff
@@ -0,0 +1,380 @@
+	dCE2CmdLiveCounter.Add(int64(b.numLive))
+	dCE2DataDeadCounter.Add(int64(b.deadMem))
+	dCE2DataLiveCounter.Add(int64(b.liveMem))
+	log.I(ctx, "DCE2: dead: %v%% %v cmds %v MB, live: %v%% %v cmds %v MB, time: %v",
+		100*b.numDead/numCmd, b.numDead, b.deadMem/1024/1024,
+		100*b.numLive/numCmd, b.numLive, b.liveMem/1024/1024,
+		dCE2Counter.Get())
+}
+
+func (b *DCEBuilder) markDependencies(nodes []NodeID) {
+	// TODO: See if a more efficient queue is necessary
+	queue := make([]NodeID, len(nodes))
+	copy(queue, nodes)
+	// nodeStr := func(nodeID NodeID) string {
+	// 	node := b.graph.GetNode(nodeID)
+	// 	if cmdNode, ok := node.(CmdNode); ok {
+	// 		return fmt.Sprintf("%v %v", cmdNode.Index, b.graph.GetCommand(api.CmdID(cmdNode.Index[0])))
+	// 	}
+	// 	return fmt.Sprintf("%v", node)
+	// }
+	for len(queue) > 0 {
+		src := queue[0]
+		queue = queue[1:]
+		// fmt.Printf("ALIVE: %s\n", nodeStr(src))
+		b.graph.ForeachDependencyFrom(src, func(tgt NodeID) error {
+			// fmt.Printf("    %s\n", nodeStr(tgt))
+			if !b.isLive[tgt] {
+				b.isLive[tgt] = true
+				queue = append(queue, tgt)
+			}
+			return nil
+		})
+	}
+}
+
+func (b *DCEBuilder) printAllCmds() {
+	b.graph.ForeachNode(func(nodeID NodeID, node Node) error {
+		alive := b.isLive[nodeID]
+		status := "ALIVE"
+		if !alive {
+			status = "DEAD "
+		}
+		if cmdNode, ok := node.(CmdNode); ok && len(cmdNode.Index) == 1 {
+			cmdID := api.CmdID(cmdNode.Index[0])
+			cmd := b.graph.GetCommand(cmdID)
+			fmt.Printf("[ %s ] (%v / %v) %v\n", status, cmdID, b.LiveCmdID(cmdID), cmd)
+		} else if obsNode, ok := node.(ObsNode); ok {
+			cmdStatus := "DEAD "
+			if obsNode.CmdID != api.CmdNoID {
+				ownerIdx := api.SubCmdIdx{uint64(obsNode.CmdID)}
+				ownerNodeID := b.graph.GetNodeID(CmdNode{ownerIdx})
+				if b.isLive[ownerNodeID] {
+					cmdStatus = "ALIVE"
+				}
+			}
+			liveCmdID := b.LiveCmdID(obsNode.CmdID)
+			fmt.Printf("[ %s ] (%v [%s] / %v) Range: %v  Pool: %v  ID: %v\n", status, obsNode.CmdID, cmdStatus, liveCmdID, obsNode.CmdObservation.Range, obsNode.CmdObservation.Pool, obsNode.CmdObservation.ID)
+		}
+		return nil
+	})
+}
+
+// func (b *DCEBuilder) printLiveCmds() {
+// 	if len(b.liveCmds) == 0 {
+// 		return
+// 	}
+// 	for i := 0; i < len(b.liveCmdIDs); i++ {
+// 		if i == len(b.liveCmdIDs)-1 || b.liveCmdIDs[i] != b.liveCmdIDs[i+1] {
+// 			fmt.Printf("  (%d / %d) %v\n", i, b.liveCmdIDs[i], b.graph.Commands[i])
+// 		}
+// 	}
+// }
+
+func (b *DCEBuilder) processLiveCmd(a arena.Arena, id api.CmdID, cmd api.Cmd) {
+	isCloned := false
+	cloneCmd := func() {
+		if !isCloned {
+			isCloned = true
+			cmd = cmd.Clone(a)
+		}
+	}
+	if b.LiveCmdID(cmd.Caller()) != cmd.Caller() {
+		cloneCmd()
+		cmd.SetCaller(b.LiveCmdID(cmd.Caller()))
+	}
+	if len(b.orphanObs) > 0 {
+		cloneCmd()
+		extras := make(api.CmdExtras, len(cmd.Extras().All()))
+		copy(extras, cmd.Extras().All())
+		obs := extras.Observations()
+		if obs == nil {
+			newReads := make([]api.CmdObservation, 0, len(b.orphanObs))
```

Comment:
I don't want to modify Observations() directly, because it is shared with the old command, in the old capture. I also want the orphan observations to come before the command's own read observations because it is possible that the command had a different, later observation of an overlapping memory range, in which case the order of the observations matters.